DOCSP-45284 DOCSP-45287 using the C++ driver from CMake and pkg-config (#88)
mongodb/docs-cpp: source/include-link.txt

.. _cpp-include-link:

===========================================
Include and Link the Driver in Your Program
===========================================

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

.. facet::
   :name: genre
   :values: reference

> 

Author: I concur with the addition of an "Overview" section. However, I think that the path placeholder is necessary. The sample program is around 150 lines long, plus it refers to another header file in the driver source tree. I suppose that if the docs site rendering will collapse the code listing down and give the reader the option of expanding it, then including the entire code listing might not be as much of an issue. If the docs site does this sort of thing, can you point to me to an example of it?

.. _cpp-include-link:

===========================================
Include and Link the Driver in Your Program
===========================================

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

.. facet::
   :name: genre
   :values: reference

Overview
--------

In this guide you will learn how to use the {+driver-short+} in your own projects. Examples are provided for using the {+driver-short+} with CMake and with ``pkg-config``.

Example Code
~~~~~~~~~~~~

> The examples on this page use the ``view_and_value.cpp`` example program from the {+driver-short+} source code. Obtain the {+driver-short+} source code by following the :ref:`cpp-quick-start-download-and-install` guide, if you have not previously done so.

Author: Nice! I didn't know that you could link to GitHub resources in this way.

You do not need to run this program on a MongoDB server.

Library References
~~~~~~~~~~~~~~~~~~

The examples use references to components of the {+driver-short+}. For CMake projects, the ``mongo::bsoncxx_shared`` target is used. For pkg-config projects, the ``libbsoncxx`` package is used. You may use the following library targets, depending on the needs of your project:

- CMake
  - ``mongo::bsoncxx_shared``
  - ``mongo::mongoccxx_shared``
  - ``mongo::bsoncxx_static``
  - ``mongo::mongoccxx_static``
- pkg-config
  - ``libbsoncxx``
  - ``libmongocxx``
  - ``libbsoncxx-static``
  - ``libmongocxx-static``

The availability of targets is dependent on the particular installation method.

In the sections that follow, replace ``<path-to-mongo-cxx-driver-sources>`` with the actual path where the {+driver-short+} source tree is located on your system.

CMake
-----

> You can use the {+driver-short+} through CMake. CMake provides the ``find_package`` command, which your project can use to locate the {+driver-short+} once it is installed. Alternatively, your project can use the advanced ``add_subdirectory`` command without requiring installation of the {+driver-short+}.

Reviewer: S: I think the first sentence can be more specific:

```suggestion
You can use CMake to include the {+driver-short+} in your project. CMake provides the ``find_package`` command, which your project can use to locate the {+driver-short+} once it is installed. Alternatively, your project can use the advanced ``add_subdirectory`` command without installing the {+driver-short+}.
```

With Driver Installation
~~~~~~~~~~~~~~~~~~~~~~~~

> Installing the {+driver-short+} enables using it in a simple and straighforward way. After installing, you can use CMake's ``find_package`` command to integrate the {+driver-short+} with your project. To learn how to install the {+driver-short+}, visit the following guides:
> 
> - :ref:`cpp-quick-start-download-and-install` to install from source
> - :ref:`cpp-installation-advanced` to install from packages

Reviewer: S: the first sentence uses some subjective language; maybe just remove it.

Also, I think a ..tip admonition would be useful for the information about visiting installation guides. We often use tips when linking out to other relevant pages.

```suggestion
After installing the {+driver-short+}, you can use CMake's ``find_package`` command to integrate the driver with your project.

.. tip::

   To learn how to install the {+driver-short+}, visit 
   the following guides:

   - :ref:`cpp-quick-start-download-and-install` to 
      install from source
   - :ref:`cpp-installation-advanced` to install from 
      packages
```

To use the ``find_package`` command, create a ``CMakeLists.txt`` file in your project directory. The following example creates a ``CMakeLists.txt`` file in the ``/home/user/project1`` project directory that uses ``find_package``:

.. literalinclude:: /includes/cmake_with_driver_installation.txt
  :caption: /home/user/project1/CMakeLists.txt
  :start-after: -- sphinx-include-start --

> Then, you can build and run your project. The execution should produce output resembling the following sample output:

Reviewer: S: Since this code example includes multiple commands and expected output, I'd break it up, and maybe omit some of the expected output. The structure could look something like:

Then, run the following commands to build your project:

code block:
cd /home/user/project1
cmake -Bbuild
cmake --build build

These commands return information about the build process. After the build completes, run the following command to run the project:

code block:
./build/view_and_value

The command line output resembles the following code:

code block:
the expected code output

   -- Detecting CXX compile features
   -- Detecting CXX compile features - done
   -- Performing Test CMAKE_HAVE_LIBC_PTHREAD
   -- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Success
   -- Found Threads: TRUE
   -- Configuring done (0.3s)
   -- Generating done (0.0s)
   -- Build files have been written to: /home/user/project1/build
   $ cmake --build build
   [ 50%] Building CXX object CMakeFiles/view_and_value.dir/<path-to-mongo-cxx-driver-sources>/examples/bsoncxx/view_and_value.cpp.o
   [100%] Linking CXX executable view_and_value
   [100%] Built target view_and_value
   $ ./build/view_and_value
   { "team" : "platforms", "id" : { "$oid" : "66f4be6fef9eb8b9240619f0" }, "members" : [ "tyler", "jason", "drew", "sam", "ernie", "john", "mark", "crystal" ] }
   Got key, key = team
   Got String!
   Got key, key = id
   Got ObjectId!
   Got key, key = members
   Got Array!
   array element: tyler
   array element: jason
   array element: drew
   array element: sam
   array element: ernie
   array element: john
   array element: mark
   array element: crystal
   as expected, we have a team
   document has 3 keys.
   document keys are:
   team
   id
   members

> If you encounter errors, you may need to specify additional options to the initial CMake command. The specific options depend on your particular environment, however there are two options that address most common issues. For instance, if your compiler does not default to at least C++17, use the ``-DCMAKE_CXX_STANDARD=17`` CMake option. If you installed the driver to a non-standard location, you may need to specify the ``-DCMAKE_PREFIX_PATH=/<path-to-mongo-cxx-driver-installation>`` option. For example:

Reviewer: Since you reference this troubleshooting info later on, I wonder if it should go in a separate "Troubleshooting" section? The Pymongo docs do something similar on a few pages, like [this one](https://www.mongodb.com/docs/languages/python/pymongo-driver/current/connect/stable-api/#troubleshooting). Maybe you could add it as a third subsection under Cmake.

I'd also reword slightly:
```suggestion
If you encounter errors, you may need to specify additional options to the initial CMake command. The specific options depend on your particular environment. However, the following options address most common issues:

- If your compiler does not default to at least C++17, use the ``-DCMAKE_CXX_STANDARD=17`` CMake option.

- If you installed the driver to a non-standard location, specify the ``-DCMAKE_PREFIX_PATH=/<path-to-mongo-cxx-driver-installation>`` option. For example:
```

.. code-block:: none

   $ cmake -Bbuild -DCMAKE_CXX_STANDARD=17 -DCMAKE_PREFIX_PATH=/opt/mongodb/cxx-driver

Without Driver Installation
~~~~~~~~~~~~~~~~~~~~~~~~~~~

> Alternatively, you can use CMake's ``add_subdirectory`` command without installing the {+driver-short+}. This is an advanced technique that, unlike the  ``find_package`` command, does not support specifying version constraints. There are additional drawbacks associated with ``add_subdirectory``, but discussion of those is beyond the scope of this document.

Reviewer: S: I don't think you need to mention the additional drawbacks if they're outside the scope
```suggestion
Alternatively, you can use CMake's ``add_subdirectory`` command without installing the {+driver-short+}. This is an advanced technique that, unlike the  ``find_package`` command, does not support specifying version constraints. 
```

.. note::

   You may need to adjust build flags for your project to
   accommodate this approach. Your project's invocation of CMake must include any flags or

>    options that you would have passed to the {+driver-short+} as part of its build.

Reviewer: S: to address the Vale issue below, maybe:
```suggestion
   options that are normally passed to the {+driver-short+} as part of its build.
```

To use the ``add_subdirectory`` command, create a ``CMakeLists.txt`` file in your project directory. The following example creates a ``CMakeLists.txt`` file in the ``/home/user/project2`` project directory that uses ``add_subdirectory``:

.. literalinclude:: /includes/cmake_without_driver_installation.txt
  :caption: /home/user/project2/CMakeLists.txt
  :start-after: -- sphinx-include-start --

.. note::

>    Note that in this approach the CMake target is referred to simply
>    as ``bsoncxx_shared``, without the ``mongo::`` namespace. The namespace

Reviewer: S: Edits for brevity/active voice:

```suggestion
   The preceding example uses the ``bsoncxx_shared`` CMake target without the ``mongo::`` namespace. The namespace
```

   is added as part of the CMake module installation, which is not performed
   in this approach.

> Then, you can build and run your project. The following shows a truncated example of the CMake output and the program execution output immediately following:

Reviewer: S: same suggestions here about breaking up the code example with multiple commands/outputs

.. code-block:: none

   $ cmake -Bbuild
   $ cmake --build build
   $ ./build/view_and_value 
   { "team" : "platforms", "id" : { "$oid" : "67207dcf532837a4470cc090" }, "members" : [ "tyler", "jason", "drew", "sam", "ernie", "john", "mark", "crystal" ] }
   Got key, key = team
   Got String!
   Got key, key = id
   Got ObjectId!
   Got key, key = members
   Got Array!
   array element: tyler
   array element: jason
   array element: drew
   array element: sam
   array element: ernie
   array element: john
   array element: mark
   array element: crystal
   as expected, we have a team
   document has 3 keys.
   document keys are: 
   team 
   id 
   members

If you encounter errors, then refer to the preceding section, which includes suggested additional CMake options for resolving common build errors.

pkg-config
----------

> If your project is not CMake-based, you can use ``pkg-config`` to integrate the {+driver-short+} with your project. Because pkg-config provides less flexibility than CMake, we recommend preferring the CMake-based approach when possible.

Reviewer: S: I think "using" is a slightly better word here than "preferring"
```suggestion
If your project is not CMake-based, you can use ``pkg-config`` to integrate the {+driver-short+} with your project. Because pkg-config provides less flexibility than CMake, we recommend using the CMake-based approach when possible.
```

You can only use the {+driver-short+} with pkg-config if you fully install the driver.

> The following code uses ``pkg-config`` to include and link the {+driver-short+}. Replace the ``/<path-to-mongo-cxx-driver-sources>`` placeholder with the location of the {+driver-short+} source tree is located on your system:

Reviewer: S:
```suggestion
The following code uses ``pkg-config`` to include and link the {+driver-short+}. Replace the ``<path-to-mongo-cxx-driver-sources>`` placeholder with the location of the {+driver-short+} source tree is located on your system:
```
